examples/generative/stylegan.py

```diff
@@ -0,0 +1,736 @@
+        )
+
+    def call(self, inputs):
+        x, noise = inputs
+        output = x + self.B * noise
+        return output
+
+
+class AdaIN(Layer):
+    def __init__(self, gain=1, **kwargs):
+        super(AdaIN, self).__init__(kwargs)
+        self.gain = gain
+
+    def build(self, input_shapes):
+        x_shape = input_shapes[0]
+        w_shape = input_shapes[1]
+
+        self.w_channels = w_shape[-1]
+        self.x_channels = x_shape[-1]
+
+        self.dense_1 = DenseBlock(self.x_channels, gain=1)
+        self.dense_2 = DenseBlock(self.x_channels, gain=1)
+
+    @tf.function(experimental_relax_shapes=True)
+    def call(self, inputs):
+        x, w = inputs
+        ys = tf.reshape(self.dense_1(w), (-1, 1, 1, self.x_channels))
+        yb = tf.reshape(self.dense_2(w), (-1, 1, 1, self.x_channels))
+
+        output = ys * x + yb
+        return output
+
+
+"""
+We create functions to build generators and discriminators at different resolution e.g.
+from 4x4, 8x8,... to 1024x1024.
+"""
+
+
+def Mapping(num_stages, input_shape=512):
+    z = Input(shape=(input_shape))
+    w = PixelNorm()(z)
+    for i in range(8):
+        w = DenseBlock(512, lrmul=0.01)(w)
+        w = LeakyReLU(0.2)(w)
+    w = tf.tile(tf.expand_dims(w, 1), (1, num_stages, 1))
+    return tf.keras.Model(z, w, name="mapping")
+
+
+def GenBlock(filter_num, res, input_shape, is_base):
+
+    input_tensor = Input(shape=input_shape, name=f"g_{res}")
+    noise = Input(shape=(res, res, 1), name=f"noise_{res}")
+    w = Input(shape=512)
+
+    x = input_tensor
+
+    if not is_base:
+        x = UpSampling2D((2, 2))(x)
+        x = ConvBlock(filter_num, 3)(x)
+
+    x = AddNoise()([x, noise])
+    x = LeakyReLU(0.2)(x)
+    x = InstanceNormalization()(x)
+    x = AdaIN()([x, w])
+
+    x = ConvBlock(filter_num, 3)(x)
+    x = AddNoise()([x, noise])
+    x = LeakyReLU(0.2)(x)
+    x = InstanceNormalization()(x)
+    x = AdaIN()([x, w])
+
+    return tf.keras.Model([input_tensor, w, noise], x, name=f"genblock_{res}x{res}")
+
+
+def DiscBase(filter_num, res):
+    return Sequential(
+        [
+            Input(shape=(res, res, filter_num), name=f"d_{res}"),
+            MinibatchStd(),
+            ConvBlock(filter_num, 3),
+            LeakyReLU(0.2),
+            Flatten(),
+            Dense(filter_num),
+            LeakyReLU(0.2),
+            Dense(1),
+        ],
+        name=f"d_{res}",
+    )
+
+
+def DiscBlock(filter_num_1, filter_num_2, res):
+    return Sequential(
+        [
+            Input(shape=(res, res, filter_num_1), name=f"d_{res}"),
+            ConvBlock(filter_num_1, 3),
+            LeakyReLU(0.2),
+            ConvBlock(filter_num_2, 3),
+            LeakyReLU(0.2),
+            AveragePooling2D((2, 2)),
+        ],
+        name=f"d_{res}",
+    )
+
+
+"""
+## Build StyleGAN with custom train step
+"""
+
+
+class Phase(Enum):
+    TRANSITION = 1
+    STABLE = 2
+
+
+class StyleGAN(tf.keras.Model):
```

Comment:
This is a very large class. Do you think it could be beneficial to readability to split it into 3, `Generator`, `Discriminator`, and `StyleGAN`?

E.g. `Generator` should have `make_blocks`, `grow_blocks`, maybe also noise-related or image generation-related methods. `StyleGAN` would have `compile` and `train_step`.

Just a thought -- you're better able than me to tell whether this would help.